extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -0,0 +1,661 @@
+                        action: 'delete',
+                        target: {
+                            owner: bucketData.owner,
+                            bucket: bucketData.bucket,
+                            key: obj.Key,
+                        },
+                    });
+                }
+                if (!alreadySent && rules.Expiration.Days &&
+                rules.Expiration.Days <= daysSinceCR) {
+                    alreadySent = true;
+                    this.sendObjectEntry({
+                        action: 'delete',
+                        target: {
+                            owner: bucketData.owner,
+                            bucket: bucketData.bucket,
+                            key: obj.Key,
+                        },
+                    });
+                }
+            }
+            // if (rules.Transitions && !alreadySent) {}
+
+            return done();
+        });
+    }
+
+    /**
+     * Handles comparing rules for non-versioned objects
+     * @param {object} bucketData - bucket data
+     * @param {string} bucketData.bucket - bucket name
+     * @param {string} bucketData.owner - owner id
+     * @param {string} [bucketData.prefix] - prefix
+     * @param {string} [bucketData.nextKeyMarker] - next key marker for
+     *   versioned buckets
+     * @param {string} [bucketData.nextVersionIdMarker] - next version id marker
+     *   for versioned buckets
+     * @param {string} [bucketData.nextContinuationToken] - next continuation
+     *   token marker for non-versioned buckets
+     * @param {array} bucketLCRules - array of bucket lifecycle rules
+     * @param {array} object - object or object version
+     * @param {Logger.newRequestLogger} log - logger object
+     * @param {function} done - callback(error, data)
+     * @return {undefined}
+     */
+    _getRules(bucketData, bucketLCRules, object, log, done) {
+        const tagParams = { Bucket: bucketData.bucket, Key: object.Key };
+        if (object.VersionId) {
+            tagParams.VersionId = object.VersionId;
+        }
+        return this.s3target.getObjectTagging(tagParams,
+        (err, tags) => {
+            if (err) {
+                log.error('failed to get tags', {
+                    method: 'LifecycleTask._compareContents',
+                    error: err,
+                    bucket: bucketData.bucket,
+                    objectKey: object.Key,
+                });
+                return done(err);
+            }
+            // tags.TagSet === [{ Key: '', Value: '' }, ...]
+            const filterRules = this._filterRules(bucketLCRules, object, tags);
+
+            // reduce filteredRules to only get earliest dates
+            return done(null, this._getApplicableRules(filterRules));
+        });
+    }
+
+    _compareContents(bucketData, lcRules, contents, log, done) {
+        if (!contents) {
+            return done();
+        }
+        return async.each(contents, (obj, cb) => {
```

Comment:
Since contents can hold a fair number of elements (up to 1000 or maybe more?) it's preferable to use `eachLimit` to limit the number of concurrent requests (say, 10 should be good enough, we can refine later).